Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs

```diff
@@ -0,0 +1,210 @@
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization;
+
+namespace Content.Server.GameObjects.Components.Chemistry
+{
+    [RegisterComponent]
+    public class SmokeComponent : Component
+    {
+        private const float ExposeDelay = 0.5f;
+
+        public override string Name => "Smoke";
+
+        [Dependency] private readonly IMapManager _mapManager = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+        private IEntity _inception;
+
+        private bool _running;
+        private float _timer;
+        private float _exposeTimer;
+
+        private int _amount;
+        private float _lifetime;
+        private float _spreadDelay;
+        private float _averageLifetime;
+
+        public bool IsInception => _inception == Owner;
+
+        public void Activate(Solution solution, int amount, float duration, float spreadDelay)
+        {
+            if (_running)
+                return;
+            TryAddSolution(solution);
+            Start(Owner, amount, duration+2*amount*spreadDelay, spreadDelay, duration+amount*spreadDelay);
+        }
+
+        public void Update(float frameTime)
+        {
+            if (!_running)
+                return;
+
+            if (IsInception)
+            {
+                if (_exposeTimer > ExposeDelay)
+                {
+                    _exposeTimer = 0;
+                    ReactWithTileAndEntities();
+                }
+                _exposeTimer += frameTime;
+            }
+            else
+            {
+                if (_inception.TryGetComponent(out SmokeComponent smokeComp) &&
+                    smokeComp._exposeTimer > ExposeDelay)
+                {
+                    ReactWithTileAndEntities();
+                }
+            }
+
+            _timer += frameTime;
+
+            if (_timer > _lifetime)
+            {
+                Owner.Delete();
+            }
+        }
+
+        private void Start(IEntity inception, int amount, float lifetime, float spreadDelay, float averageLifetime)
+        {
+            _inception = inception;
+            _amount = amount;
+            _lifetime = lifetime;
+            _spreadDelay = spreadDelay;
+            _averageLifetime = averageLifetime;
+            _running = true;
+
+            Owner.SpawnTimer((int) (_spreadDelay*1000), SpreadSmoke);
+        }
+
+        private void TryAddSolution(Solution solution)
+        {
+            if (solution.TotalVolume == 0)
+                return;
+
+            if (!Owner.TryGetComponent(out SolutionContainerComponent contents))
+                return;
+
+            var addSolution = solution.SplitSolution(ReagentUnit.Min(solution.TotalVolume,contents.EmptyVolume));
+
+            var result = contents.TryAddSolution(addSolution);
+
+            if (!result)
+                return;
+
+            if (Owner.TryGetComponent(out SpriteComponent sprite))
+                sprite.Color = contents.SubstanceColor;
```

Comment:
Is this expensive? Using sprite states pushes all of the data over the network and doing this for a lot of entities could be a lot of bandwidth.